### PR TITLE
txn: try to commit pessimistic txn with concurrent ddl

### PR DESCRIPTION
**Pull request by cfzjywxk**
### What problem does this PR solve?

Issue Number:  #17932

Problem Summary:
Try to commit pessimistic transactions with related table schema changes by concurrent DDLs

### What is changed and how it works?

Proposal: #17932 
This is the first part

What's Changed:

1. Try to record `ActionType` for each schema change, return both the changed table id array, as well as the changed action type array.(Maybe some schema change diff merge could be done by their type relation)
2. Use the changed type array to check if this schema change for the current pessimistic transaction is amendable. If so, try to amend the transaction memory buffer based on the schema change type, and then try to commit the pessimistic transaction as usual but not report `ErrSchemaChanged`.
3. Try to implement `add column` and `drop column` for the amend logic by now. No memory buffer amendation is needed for these two types.

How it Works:

### Check List 

Tests 

- Unit test
- Manual test (add detailed scripts or steps below)

Side effects

### Release note 

-

### CODE REVIEW

File: domain/domain.go [hunk 1]
```diff
@@ -80,19 +80,19 @@ type Domain struct {
 // loadInfoSchema loads infoschema at startTS into handle, usedSchemaVersion is the currently used
 // infoschema version, if it is the same as the schema version at startTS, we don't need to reload again.
 // It returns the latest schema version, the changed table IDs, whether it's a full load and an error.
-func (do *Domain) loadInfoSchema(handle *infoschema.Handle, usedSchemaVersion int64, startTS uint64) (int64, []int64, bool, error) {
+func (do *Domain) loadInfoSchema(handle *infoschema.Handle, usedSchemaVersion int64, startTS uint64) (int64, []int64, []uint64, bool, error) {
```
Comment (coocood) [inline domain/domain.go:83]: There are too many returned values, we better make them named.

File: store/tikv/2pc.go [hunk 2]
```diff
@@ -1374,15 +1375,50 @@ func (c *twoPhaseCommitter) stripNoNeedCommitKeys() {
 	c.mutations = m.subRange(0, newIdx)
 }
 
+const amendableType = uint64((1 << model.ActionAddColumn) | (1 << model.ActionDropColumn))
```
Comment (tiancaiamao) [inline store/tikv/2pc.go:1378]: What happen if there are two DDL changes that drop the column and than add back the column?

File: domain/schema_validator.go [hunk 3]
```diff
@@ -137,44 +139,58 @@ func (s *schemaValidator) Update(leaseGrantTS uint64, oldVer, currVer int64, cha
...
+
+	changedTblMap := make(map[int64]uint64)
 	for _, item := range newerDeltas {
-		if hasRelatedTableID(item.relatedIDs, tableIDs) {
-			return true
+		for i, tblID := range item.relatedIDs {
+			for _, relatedTblID := range tableIDs {
+				if tblID == relatedTblID {
+					if !getAllInfo {
+						return nil, nil, true
```
Comment (tiancaiamao) [inline domain/schema_validator.go:171]: If we change the Delta to table based, the for loop

```
for ...
   for ...
      for
``` 

could be terminated early once the table ID does not match.

Using current data struct we have to exhaust the loop.

File: domain/schema_validator.go [hunk 4]
```diff
@@ -137,44 +139,58 @@ func (s *schemaValidator) Update(leaseGrantTS uint64, oldVer, currVer int64, cha
...
 	if len(newerDeltas) == len(s.deltaSchemaInfos) {
 		metrics.LoadSchemaCounter.WithLabelValues(metrics.SchemaValidatorCacheMiss).Inc()
 		logutil.BgLogger().Info("the schema version is much older than the latest version", zap.Int64("currVer", currVer),
 			zap.Int64("latestSchemaVer", s.latestSchemaVer), zap.Reflect("deltas", newerDeltas))
-		return true
+		return nil, nil, true
 	}
+
+	changedTblMap := make(map[int64]uint64)
 	for _, item := range newerDeltas {
```
Comment (tiancaiamao) [inline domain/schema_validator.go:166]: Currently there are only 100 schema delta are saved, if a transaction last too long and more than 100 schema changes happen during that time, the transaction will fail.

File: domain/domain.go [hunk 5]
```diff
@@ -233,39 +233,43 @@ func isTooOldSchema(usedVersion, newVersion int64) bool {
...
 		if err != nil {
-			return false, nil, err
+			return false, nil, nil, err
 		}
 		if canSkipSchemaCheckerDDL(diff.Type) {
 			continue
 		}
 		tblIDs = append(tblIDs, ids...)
+		for i := 0; i < len(ids); i++ {
+			actions = append(actions, uint64(1<<diff.Type))
```
Comment (coocood) [inline domain/domain.go:268]: Why not just append the action itself?

Comment (coocood) [reply domain/domain.go:268]: What if the action exceed 63 in the future?

File: domain/domain.go [hunk 6]
```diff
@@ -233,39 +233,43 @@ func isTooOldSchema(usedVersion, newVersion int64) bool {
 // Return true if the schema is loaded successfully.
 // Return false if the schema can not be loaded by schema diff, then we need to do full load.
 // The second returned value is the delta updated table and partition IDs.
-func (do *Domain) tryLoadSchemaDiffs(m *meta.Meta, usedVersion, newVersion int64) (bool, []int64, error) {
+func (do *Domain) tryLoadSchemaDiffs(m *meta.Meta, usedVersion, newVersion int64) (bool, []int64, []uint64, error) {
```
Comment (coocood) [inline domain/domain.go:236]: How about returning the `diffs` directly instead of two slices?

File: domain/schema_validator.go (hunk 4 above)
Comment (coocood) [reply domain/schema_validator.go:166]: This is ensured by L166.

Comment (coocood) [reply domain/schema_validator.go:166]: if len(newerDeltas) == len(s.deltaSchemaInfos) {

File: store/tikv/2pc.go (hunk 2 above)
Comment (cfzjywxk) [reply store/tikv/2pc.go:1378]: It's ok the newly added column will use the default value

**Pull request closed without merging**